src/checkpoint/traits/user_traits.h

```diff
@@ -0,0 +1,108 @@
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_USER_TRAITS_CHECKPOINT_TRAITS_H
+#define INCLUDED_USER_TRAITS_CHECKPOINT_TRAITS_H
+
+namespace checkpoint {
+
+namespace {
+  //Pass around a list of types via storage in the parameters of a tuple
+  template<typename First, typename... Second>
+  struct addToTuple {};
+
+  template<typename First, typename... Second>
+  struct addToTuple<First, std::tuple<Second...>> {
+      using type = std::tuple<First, Second...>;
+  };
+  
+  //Take types from a tuple and template our Serializer type on those.
+  template<template<typename...> typename SerializerT, typename... UserTraits>
+  struct fromTuple {};
+
+  template<template<typename...> typename SerializerT, typename... UserTraits> 
+  struct fromTuple<SerializerT, std::tuple<UserTraits...>> {
+      using type = SerializerT<UserTraits...>;
+  };
+
+  
+  //Get a SerializerT with all the same template parameters except all instances of a given type.
+  //No change if ToRemove is not present in UserTraits.
+  template<typename ToRemove, typename... UserTraits>
+  struct removeTraitImpl {
+     using Traits = std::tuple<UserTraits...>;
+  };
+
+  template<typename ToRemove, typename... UserTraits>
+  struct removeTraitImpl<ToRemove, ToRemove, UserTraits...> {
+      using Traits = typename removeTraitImpl<ToRemove, UserTraits...>::Traits;
+  };
+
+  template<typename ToRemove, typename Other, typename... UserTraits>
+  struct removeTraitImpl<ToRemove, Other, UserTraits...>{
+      using Traits = typename addToTuple<Other, typename removeTraitImpl<ToRemove, UserTraits...>::Traits>::type;
+  };
+
+  template<typename ToRemove, template<typename...> typename SerializerT, typename... UserTraits>
+  struct removeTrait {
+      using type = typename fromTuple<SerializerT, typename removeTraitImpl<ToRemove, UserTraits...>::Traits>::type;
+  };
+}
+
+namespace SerializerUserTraits {
+  template<typename Trait, template<typename... UserTraits> typename SerializerT, typename... UserTraits>
+  using hasTrait = std::disjunction<std::is_same<Trait, UserTraits>...>;
+
+  template<typename Trait, template<typename...> typename SerializerT, typename... UserTraits>
+  auto& withoutTrait(SerializerT<UserTraits...>& obj){
+    return reinterpret_cast<typename removeTrait<Trait, SerializerT, UserTraits...>::type&>(obj);
```

Comment:
There's some difficulty since the serializers are expected to be used by the same reference. I've been looking into it, and technically any mechanism we use to refer to the same memory space by the two different types is undefined. The user traits aren't used to change any members, so we should be good on memory positioning, but strict aliasing rules could cause issues w.r.t. compilers keeping member variables in different registers. 

I haven't tested, but I think we ensure correct reads/writes by:
1. Require all serializers be stored as volatile variables, or all non-reference/non-pointer members of serializers be declared volatile
2. Require all serializer types declared with the __may_alias__ trait, which should be available on gcc/clang/icc. I expect this would have more side effects on compiler optimizations.

Any thoughts?